File: mobile/lib/core/formatters/cents_formatter.dart
/// Format server-provided integer cents without floating-point arithmetic.
String formatCents(int cents) {
  final negative = cents < 0;
  final magnitude = negative ? -cents : cents;
  final whole = magnitude ~/ 100;
  final fraction = magnitude % 100;
  final wholeText = _groupThousands(whole.toString());
  final fractionText = fraction.toString().padLeft(2, '0');
  final sign = negative && magnitude != 0 ? '-' : '';

  return '${sign}Bs. $wholeText.$fractionText';
}

String _groupThousands(String digits) {
  final groups = <String>[];
  for (var end = digits.length; end > 0;) {
    final start = end > 3 ? end - 3 : 0;
    groups.add(digits.substring(start, end));
    end = start;
  }
  return groups.reversed.join(',');
}

String formatBolivianos(int cents) => formatCents(cents);

File: web/src/core/cents-formatter.ts
export type CentsInput = number | string | bigint;

function integerText(cents: CentsInput): string {
  if (typeof cents === "number") {
    if (!Number.isSafeInteger(cents)) {
      throw new TypeError("Cents must be a safe integer.");
    }
    return String(cents);
  }

  const text = typeof cents === "bigint" ? cents.toString() : cents;
  if (!/^-?\d+$/.test(text)) {
    throw new TypeError("Cents must be an integer.");
  }
  return text;
}

function groupThousands(digits: string): string {
  return digits.replace(/\B(?=(\d{3})+(?!\d))/g, ".");
}

/** Format server-provided integer cents without floating-point arithmetic. */
export function formatCents(cents: CentsInput): string {
  const raw = integerText(cents);
  const negative = raw.startsWith("-");
  const digits = (negative ? raw.slice(1) : raw).replace(/^0+(?=\d)/, "");
  const whole = digits.length > 2 ? digits.slice(0, -2) : "0";
  const fraction = digits.length > 1 ? digits.slice(-2) : `0${digits}`;
  const sign = negative && digits !== "0" ? "-" : "";

  return `${sign}Bs. ${groupThousands(whole)},${fraction}`;
}

export const formatBolivianos = formatCents;

/** Format a balance, adding an explicit plus sign only for positive values. */
export function formatSignedCents(cents: CentsInput): string {
  const raw = integerText(cents);
  const formatted = formatCents(raw);
  return raw.startsWith("-") || /^0+$/.test(raw) ? formatted : `+${formatted}`;
}
